### ISSUE TITLE
[BUG]: Segmentation fault running script calling into python using `mojo run` (due to mismatch between libpython that mojo finds vs. what python packages with native modules were built against)

### ISSUE THREAD

**Issue by cjfuller** [modularml/mojo#673]
### Bug description

I got a segmentation fault while running a script that calls into python with `mojo run` and it said to post the information here as a bug report.

```
$ mojo run py.mojo
[115788:115788:20230909,110056.735000:ERROR elf_dynamic_array_reader.h:64] tag not found
[115788:115788:20230909,110056.735051:ERROR elf_dynamic_array_reader.h:64] tag not found
[115788:115788:20230909,110056.735074:ERROR elf_dynamic_array_reader.h:64] tag not found
[115788:115788:20230909,110056.735085:ERROR elf_dynamic_array_reader.h:64] tag not found
[115788:115788:20230909,110056.735113:ERROR elf_dynamic_array_reader.h:64] tag not found
[115788:115788:20230909,110056.736069:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq: No such file or directory (2)
[115788:115788:20230909,110056.736087:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: No such file or directory (2)
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.      Program arguments: mojo run py.mojo
#0 0x000055b5e4f86957 (/home/colin/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5bb957)
#1 0x000055b5e4f8452e (/home/colin/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5b952e)
#2 0x000055b5e4f8702f (/home/colin/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5bc02f)
#3 0x00007fb6f104c520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
#4 0x00007fb6e44710ec PyUnicode_AsUTF8AndSize (/home/linuxbrew/.linuxbrew/Cellar/python@3.11/3.11.4/lib/libpython3.11.so.1.0+0x1f60ec)
#5 0x00007fb654003234
zsh: segmentation fault  mojo run py.mojo
```

You'll note in the dump above, it's referencing a libpython3.11 that's installed on my system, but below (see "system information") when I created the virutalenv for the python packages, I got python3.8, also installed on my system.

If I create the virutalenv with 3.11 explicitly via `python3.11 -m venv .venv` instead, then there is no segfault, and it works normally.

(Could be related to #642 ?)

### Steps to reproduce

Here's the files I'm using:

py.mojo:
```
from python import Python

def main():
    let pd = Python.import_module("pandas")
    print(pd.read_csv("/home/colin/numbers.csv"))
```
numbers.csv was generated with:
```shell
for i in `seq 0 2499`
do
echo $i >> ~/numbers.csv
done
```
the size and exact file contents don't really matter here; (`echo "1" > ~/numbers.csv` also has the problem)

(If I run the equivalent in a python repl, then it works normally and creates the dataframe.)

### System information

```shell
WSL 2 (Win 11 Pro 22H2, build 22621.2215) / Ubuntu 22.04.2 LTS

in powershell:

> wsl --version
WSL version: 1.2.5.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22621.2215
```

in WSL/ubuntu:
```
$ uname -a
Linux eudyptes 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
$ mojo --version
mojo 0.2.1 (64d14e85)
```

This is running in a python virtualenv that I created via:
```
$ python --version
Python 3.8.17
$ python -m venv .venv
$ source .venv/bin/activate
$ pip install pandas
$ pip list
Package         Version
--------------- ------------
numpy           1.24.4
pandas          2.0.3
pip             23.0.1
python-dateutil 2.8.2
pytz            2023.3.post1
setuptools      56.0.0
six             1.16.0
tzdata          2023.3
```
```

**Comment by austinorr** [comment]
I use virtual envs and ran into this issue. Maybe this other https://github.com/modularml/mojo/issues/551#issuecomment-1712574082 will help you too.

**Comment by austinorr** [comment]
I think the MRE is even smaller than your example that includes loading the csv. The following also results in errors even when the MOJO_PYTHON_LIBRARY is set (so maybe not related to #642?):

I'm in a conda env with the MOJO_PYTHON_LIBRARY path set to ~/miniconda3/envs/mojo/lib/libpython3.11.so
and I have `numpy` , `matplotlib` and `pandas` installed.

this works:
```python
from python import Python


fn load() raises:
    let np= Python.import_module("numpy")


fn main() raises:
    load()
```

this fails:
```python
from python import Python


fn load() raises:
    let pd = Python.import_module("pandas")


fn main() raises:
    load()
```

and this fails:
this fails:
```python
from python import Python


fn load() raises:
    let plt= Python.import_module("matplotlib.pyplot")


fn main() raises:
    load()
```

makes it hard to follow along with the mandelbrot example which demonstrates how to do this with something like:
```python
from python import Python

var np: PythonObject = None
var plt: PythonObject = None
var colors: PythonObject = None


fn load() raises:

    np = Python.import_module("numpy")
    plt = Python.import_module("matplotlib.pyplot")
    colors = Python.import_module("matplotlib.colors")


fn main() raises:
    load()
```

with the exception of the import `numpy` example, all of these raise
```
Unhandled exception caught during execution: An error occurred in Python.
mojo: error: execution exited with a non-zero result: 1
```

**Comment by austinorr** [comment]
@cjfuller I’m glad it’s working for you now. Anything more you can provide about the ‘setting up venv differently’ ?

I agree that my setup could just be an issue with conda env support, perhaps with the way the paths to dependencies are resolved (especially if there are built binary deps like with pd and mpl).

**Comment by cjfuller** [comment]
> Anything more you can provide about the ‘setting up venv differently’ ?

I haven't experimented much here. It worked if I just made sure to invoke the `python -m venv .venv` to create the virutalenv with the same python that was in the segfault message (i.e. the python that `mojo run` was finding automatically).

**Comment by stumpOS** [comment]
Please see instructions on how to configure python [here](https://github.com/modularml/mojo/issues/537#issuecomment-1718038049). You can verify that the search space matches the location of the package by printing the paths. For example:
```
from python import Python


fn load() raises:
    try:
       let pd = Python.import_module("pandas")
    except e:
       print(e.value)
       let sys = Python.import_module("sys")
       var paths = sys.path
       for p in paths:
           print(p)


fn main() raises:
    load()
```
Note that we are currently working on updating the error message so it is more informative.
